lib/internal/modules/esm/module_job.js

```diff
@@ -143,7 +143,16 @@ class ModuleJob {
     await this.instantiate();
     const timeout = -1;
     const breakOnSigint = false;
-    await this.module.evaluate(timeout, breakOnSigint);
+    try {
+      await this.module.evaluate(timeout, breakOnSigint);
+    } catch (err) {
+      if (err.message.includes('is not defined') && (err.stack.includes("esm/loader"))) {
+        err.message = 'You cannot require from an ESM module. Replace require with import or leverage createRequire.';
+        err.stack = err.stack.replace(/ReferenceError:.*\n/,
+                                      `ReferenceError: ${err.message}\n`);
```

Comment:
@MylesBorins, I pushed another update. I am making two naive checks below:

1. If the error includes the reference error text. I could probably just check for "ReferenceError" instead, but I have not checked to see if there are other types of reference errors yet. I was hoping to write test cases while figuring that out.
2. If the stack trace includes esm/loader. Also not sure if this is necessary with the first check, but this would give me confidence that the error is due to esm module being used with a cjs require. Again, probably could write some more tests to narrow down if this is needed.

Reply:
any attempt to use an undefined variable is going to throw a reference error here, and will have esm/loader in the stack. So I think the current checks will unfortunately have a lot of false positives. Some checks you could do instead could include

* checking to see if the error is a reference error

try { test } catch (e) { console.log(e instanceof ReferenceError) }

* checking the specific error message

try { test } catch (e) { console.log(e.message.includes('test is not defined'))

Alternatively you could attempt to approach this similarly to how I did this in https://github.com/nodejs/node/commit/861a54c0e1265e5020348892b87f3ec8935dd9fe

I simply did a check for the exact error message.